Wanna merge with mrc

mloviska opened the pull request:
Test

cfconrad committed:
t/22-svirt.t: Add tests for console::sshVirtsh::add_disk() method

mloviska committed:
Rework add drives mechanism in `svirt`

SLEM 5.1 builds bootable, ready to use images for z/KVM. To make sure
that images can be expanded we need to add additional functionality to
`add_disk` to create a backingfile of given size.
For `svirt` backend ( except of hyperv ), we should be able to create
drives in following modes.

1. Create empty drive
2. Create a backing image file of given size
3. Re-use given image
4. Add cd-rom drive with given iso image

mloviska committed:
Do not override bus and dev types for xen

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+
+sub gen_kvm_img {
+    my ($self, $args) = @_;
+    # create [--object OBJECTDEF] [-q] [-f FMT] [-b BACKING_FILE] [-F BACKING_FMT] [-u] [-o OPTIONS] FILENAME [SIZE]
```
I would add input validation like:
```suggestion
    die('Missing argument size or invalid format') unless(defined(to_byte($args->{size})));
    die('Missing argument `new_image_file`') unless(defined($args->{new_image_file}));
    $args->{backingfile} //= 0; # default is __no__ backing file
    
    # create [--object OBJECTDEF] [-q] [-f FMT] [-b BACKING_FILE] [-F BACKING_FMT] [-u] [-o OPTIONS] FILENAME [SIZE]
```

mloviska replied:
Isn't the mandatory only new file parameter ?

cfconrad replied:
Didn't saw a place where you set a default size in `get_kvm_img()`. And the line also check if `to_byte()` can parse the input.

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+}
+
+sub gen_kvm_img {
+    my ($self, $args) = @_;
```
In general I like such pattern more
```suggestion
    my ($self, %args) = @_;
```

With this you would call the method like:
```perl
$self->gen_kvm_img(new_image_file => 'foo.bla', size => '10G'); 
```
instead of 
```perl
$self->gen_kvm_img({new_image_file => 'foo.bla', size => '10G'}); 
```
but you need to adopt the acces of `$args` in that method if you take it.

mloviska replied:
oh, my fault. I am on purpose building the hash, thus there is only one call of `gen_kvm_img`

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
-            my $retval = $self->run_cmd($cmd, domain => 'sshVMwareServer');
-            die "Can't create VMware image $vmware_disk_path" if $retval;
+    #in case of invalid format return error
+    if ($sp !~ /^(?<num>\d+)(?<unit>[bkKMGT])?$/) {
```
```suggestion
    if (!defined($sp) || $sp !~ /^(?<num>\d+)(?<unit>[bkKMGT])?$/) {
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
-            my $retval = $self->run_cmd($cmd, domain => 'sshVMwareServer');
-            die "Can't create VMware image $vmware_disk_path" if $retval;
+    #in case of invalid format return error
+    if ($sp !~ /^(?<num>\d+)(?<unit>[bkKMGT])?$/) {
+        return undef;
+    } else {
+        return $+{num} * $size_multiplier->{$+{unit} // 'b'};
+    }
```
```suggestion
    if (!defined($sp) || $sp !~ /^(?<num>\d+)(?<unit>[bkKMGT])?$/) {
       return undef;
     }
       
     return $+{num} * $size_multiplier->{$+{unit} // 'b'};
    
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+sub add_disk {
+    my ($self, $args) = @_;
+    my $size = $args->{size} // '20';
+    if (looks_like_number($size)) {
+        $size .= 'G';
+    }
```
Same here, we allow `5.5`.
```suggestion
    $size .= 'G' if ($size =~ /^\d+$/);
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+
+sub add_disk {
+    my ($self, $args) = @_;
+    my $size = $args->{size} // '20';
```
```suggestion
    my $size = $args->{size} // '20G';
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+    if (looks_like_number($size)) {
+        $size .= 'G';
+    }
+
```
I would add input validation that only one of `backingfile`, `create` or `cdrom` is set:
```suggestion
die('Use one out of `backingfile`, `create` or `cdrom`') if ($args->{backingfile} // 0 + $args->{cdrom} // 0 + $args->{create} // 0 != 1);
```
Of course, a 'type' or so would fit more, but I guess we do not want to chage the interface!

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+            $file = $self->download_image_file($args->{file}, $basedir);
+        } else {
+            my $qemu_args = {new_image_file => $file};
+            unless ($args->{create}) {
+                $original_image_synced = $self->download_image_file($args->{file}, $basedir);
+            }
+            $qemu_args->{backingfile} = $original_image_synced if $args->{backingfile};
```
```suggestion
            if ($args->{backingfile}) {
                $qemu_args->{backingfile} = $self->download_image_file($args->{file}, $basedir);
            }
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,126 @@ sub add_interface {
+}
+
+sub download_image_file {
+    my ($self, $src, $dest) = @_;
```
I would add here some description... That the destination must be given as path or so...

Martchus committed:
Allow one to easily configure a build with commonly useful warnings

* This change does *not* change anything unless one adds
  `-DENABLE_WARNINGS` or `-DTREAT_WARNINGS_AS_ERRORS` to the CMake
  arguments.
* At this point this is mainly intended to configure a development build
  with useful warnings enabled more easily. In other context (e.g. during
  rpm package builds) one can still configure warnings as they make sense
  in this context (e.g. by using certain rpm macros).
* If all warnings are fixed, it would make sense to enable both settings
  for CI builds.

Martchus committed:
Disable certain warnings in the XS file as it is generated by xsubpp

We cannot fix the warnings in this generated file and they look harmless.

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +326,135 @@ sub add_interface {
-            $file = ($self->vmm_family eq 'vmware' ? '' : $basedir) . $file_basename;
+    } else {
+        ### XEN or other KVM based hypervisors
+        $basedir = '/var/lib/libvirt/images/';
+        $file    = sprintf("%s%s%s.img", $basedir, $self->{name}, $args->{dev_id});
+        if ($args->{cdrom}) {
+            $file = $self->download_image_file($args->{file}, $basedir);
+        } else {
+            my $qemu_args = {new_image_file => $file};
+            if ($args->{backingfile}) {
+                $qemu_args->{backingfile} = $self->download_image_file($args->{file}, $basedir);
+            }
+            $qemu_args->{size} = $size;
+            $self->gen_kvm_img($qemu_args);
```
```suggestion
        $basedir = '/var/lib/libvirt/images/';
        $file    = sprintf("%s%s%s.img", $basedir, $self->{name}, $args->{dev_id});
        if ($args->{cdrom}) {
            $file = $self->download_image_file($args->{file}, $basedir);
        } else {
            my %$qemu_args = (new_image_file => $file);
            if ($args->{backingfile}) {
                $qemu_args{backingfile} = $self->download_image_file($args->{file}, $basedir);
            }
            $qemu_args{size} = $size;
            $self->gen_kvm_img(%qemu_args);
```

aylen384 committed:
Merge pull request #1740 from Martchus/debug-segfault

Allow one to easily configure a build with commonly useful warnings

mloviska commented on consoles/sshVirtsh.pm:
```diff
@@ -28,6 +28,7 @@ use File::Temp 'tempfile';
 use File::Basename;
 use Class::Accessor 'antlers';
 use Mojo::JSON qw(decode_json);
+use Scalar::Util qw(looks_like_number);
```
Reminder, not used

cfconrad committed:
testapi::compat_args: Fix if given value contains invalid regex

It was discovered during a call like:
```
$instance->run_ssh_command(q(rpm -q --qf '%{VERSION}\\n' ltp));
```
while `publiccloud::instance::run_ssh_command()` used `compat_args`
like:
```
testapi::compat_args({cmd => undef}, ['cmd'], @_);
```
which makes the first argument to `cmd` if not given via as hash (e.g.
`cmd => 'exit 1'`).

But this failed with
```
  Test died: Unescaped left brace in regex is illegal here in regex; marked by <-- HERE in m/^rpm -q --qf '%{ <-- HERE VERSION}\n' ltp$/ at /usr/lib/os-autoinst/testapi.pm line 2311.
  testapi::compat_args(undef, undef, "rpm -q --qf '%{VERSION}\\n' ltp") called at sle/lib/publiccloud/instance.pm line 51
```
because we use the given argument in a regex instead of do a `eq`
compare.

mergify[bot] committed:
Merge pull request #1742 from cfconrad/fix_compat_args_invalid_regex

testapi::compat_args: Fix if given value contains invalid regex

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+}
+
+sub decompress_cmd {
+    return sprintf('xz --decompress --keep --verbose --no-warn --quiet %s', shift);
```
If you wanna be strict, you might enclose the name in single ticks
```suggestion
    return sprintf(q(xz --decompress --keep --verbose --no-warn --quiet '%s'), shift);
```

Q2: 
We used `--keep` before, but do we really need it? If not, I wouldn't skip it within this PR, so it is a theoretical question ;)

mloviska replied:
Isn't it there to avoid constant downloading of the asset? Good question, maybe we should rethink this part

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+
 sub add_disk {
     my ($self, $args) = @_;
+    die('Use one out of `backingfile`, `create` or `cdrom`') if ((($args->{backingfile} // 0) + ($args->{cdrom} // 0) + ($args->{create} // 0)) != 1);
```
```suggestion
    die('Use one out of `backingfile => 1`, `create => 1` or `cdrom => 1`') if ((($args->{backingfile} // 0) + ($args->{cdrom} // 0) + ($args->{create} // 0)) != 1);
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+
+    if ($self->vmm_family eq 'vmware') {
+        my $basedir = "/vmfs/volumes/$vmware_datastore/openQA/";
+        $file = sprintf("%s%s.vmdk", $self->{name}, $args->{dev_id});
```
`$args->{dev_id}` can be `undef` at this point. Maybe we need some input validation for it as well?

mloviska replied:
Seems like another mandatory argument that has to be passed :)
I am briefly checking the callers, seems like it is always there.

mloviska replied:
and we would create a problem over here...

```
/lib/bootloader_setup.pm:            $svirt->add_disk({file => $patched_img, dev_id => 'a'});
```

cfconrad replied:
ah because we miss `create` `cdrom` or `backingfile`... what was the default before?

mloviska replied:
> ah because we miss `create` `cdrom` or `backingfile`... what was the default before?

no, we haven't missed anything. That caller comes from test repo, it is actually the only one that I could have found that wants to use the original image :/

mloviska replied:
As I understand this tree https://openqa.suse.de/tests/6901438#dependencies, they use single image without backing file.

mloviska replied:
> As I understand this tree https://openqa.suse.de/tests/6901438#dependencies, they use single image without backing file.

I am wrong, `backend::baseclass::run_ssh_cmd(cmd="qemu-img create '/var/lib/libvirt/images/openQA-SUT-6a.img' -f qcow2 -b '/var/lib/libvirt/images//SLES-15-SP3-s390x-mru-install-minimal-with-addons-Build20210820-1-Server-DVD-Updates-s390x-kvm-sle12.qcow2' 42949672960"` Sorry for the noise

mloviska replied:
I think we would break this part https://openqa.suse.de/tests/6892091

cfconrad replied:
I'm still not sure what we did wrong?

So my advice for the input validation was wrong, or?
And we need to copy the file if not `backing|create|cdrom` is given.

cfconrad replied:
We also miss a test-case for that

mloviska replied:
> I'm still not sure what we did wrong?
> 
> So my advice for the input validation was wrong, or?
> And we need to copy the file if not `backing|create|cdrom` is given.

We have missed a valid migration scenario. They boot older sle, patch it, reboot so they initiate YaST2 installer to upgrade the system. 

So at the very beginning, they download image, create backing file for it. And at the second execution of `bootloader_zkvm` they just use the XML part of the code

cfconrad replied:
who does the cleanup of the backing files?

mloviska replied:
I am aware of only this one https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/installation/bootloader_svirt.pm#L69

mloviska replied:
Not sure whether there is any clean up for xen or zkvm

cfconrad replied:
ok

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+              'done;' .
+              'else ' .
+              "cp /vmfs/volumes/$vmware_nfs_datastore/hdd/$base_orig_file $basedir;" .
+              'fi;';
```
Do you know if `$self->run_cmd()` support real multiline commands which could be specified like this?

```suggestion
            my $cmd =<<"EOT";
if test -e $abs_path_file; then 
    while lsof | grep 'cp.*$base_orig_file'; do 
        echo 'File $base_orig_file is being copied by other process, sleeping for 60 seconds'
        sleep 60
    done
else
    cp '/vmfs/volumes/$vmware_nfs_datastore/hdd/$base_orig_file' '$basedir';
fi
EOT
```

mloviska replied:
tests are passing :)

cfconrad replied:
cool

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+              'fi;';
+            $self->run_cmd($cmd, domain => 'sshVMwareServer') && die "Can't copy VMware image $file";
+            if ($args->{file} =~ /\.xz$/) {
+                $self->run_cmd(decompress_cmd($abs_path_file), domain => 'sshVMwareServer') && die "Image $abs_path_file decompression has failed!";
```
Don't we need to remove the `xz` from the name, something like:
```suggestion
                $self->run_cmd(decompress_cmd($abs_path_file), domain => 'sshVMwareServer') && die "Image $abs_path_file decompression has failed!";
                $abs_path_file =~ s/\.xz$//;
```

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+    my $vmware_datastore = get_var('VMWARE_DATASTORE', '');
+
+    if ($self->vmm_family eq 'vmware') {
+        my $basedir = "/vmfs/volumes/$vmware_datastore/openQA/";
```
```suggestion
        my $basedir = "/vmfs/volumes/$vmware_datastore/openQA/";
        my $nfs_basedir = '/vmfs/volumes/'. get_required_var('VMWARE_NFS_DATASTORE') . '/';
```
and use it below.

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,159 @@ sub add_interface {
+        $basedir = '/var/lib/libvirt/images/';
+        $file    = sprintf("%s%s%s.img", $basedir, $self->{name}, $args->{dev_id});
+        if ($args->{cdrom}) {
+            $file = $self->download_image_file($args->{file}, $basedir);
```
`download_image_file()` pretend something else then simply doing a copy via `rsync`

cfconrad commented on t/22-svirt.t:
```diff
@@ -234,4 +236,479 @@ subtest 'Method backend::svirt::open_serial_console_via_ssh()' => sub {
+            $cmd = shift @last_ssh_commands;
+            like($cmd, qr/vmkfstools -v1 --deletevirtualdisk $exp_fullpath/, "Check name of image to be deleted");
+            $cmd = shift @last_ssh_commands;
+            like($cmd, qr%cp\s+/vmfs/volumes/nfs/hdd/$original_filename\s+$vmware_openqa_datastore*;%, "Copy hdd to $vmware_openqa_datastore");
```
```suggestion
            like($cmd, qr%cp\s+/vmfs/volumes/nfs/hdd/$original_filename\s+$vmware_openqa_datastore\*;%, "Copy hdd to $vmware_openqa_datastore");
```

mloviska replied:
hm...I guess it wouldn't match the end anymore, or ?

cfconrad commented on t/22-svirt.t:
```diff
@@ -298,12 +302,56 @@ subtest 'Method consoles::sshVirtsh::add_disk()' => sub {
+            $cmd = shift @last_ssh_commands;
+            like($cmd, qr%xz --decompress --keep --verbose --no-warn --quiet '$exp_fullpath_xz'%, 'Uncompress file with unxz');
+            $cmd = shift @last_ssh_commands;
+            like($cmd, qr/vmkfstools -v1 --clonevirtualdisk $vmware_openqa_datastore$file --diskformat thin $exp_fullpath/, "Check size");
```
I guess we need to use `$file` without `.xz` extention.

mloviska replied:
yes, already fixed :D

cfconrad commented on t/22-svirt.t:
```diff
@@ -469,10 +525,10 @@ subtest 'Method consoles::sshVirtsh::add_disk()' => sub {
-   #$target_nodelist = $svirt->{domainxml}->findnodes('domain/devices/disk[@type="file" and @device="cdrom"]/target[@dev="sd' . $dev_id . '" and @bus="scsi"]');
-            my $target_nodelist = $svirt->{domainxml}->findnodes('domain/devices/disk[@type="file" and @device="cdrom"]/target[@dev="xvd' . $dev_id . '" and @bus="xen"]'); # This looks buggy! I think we miss the `else`
+            my $target_nodelist = $svirt->{domainxml}->findnodes('domain/devices/disk[@type="file" and @device="cdrom"]/target[@dev="sd' . $dev_id . '" and @bus="scsi"]');
+#            my $target_nodelist = $svirt->{domainxml}->findnodes('domain/devices/disk[@type="file" and @device="cdrom"]/target[@dev="xvd' . $dev_id . '" and @bus="xen"]'); # This looks buggy! I think we miss the `else`
```
lets remove the comment

cfconrad commented on t/22-svirt.t:
```diff
@@ -569,8 +624,8 @@ subtest 'Method consoles::sshVirtsh::add_disk()' => sub {
             });
-            like($last_ssh_commands[0], qr%^rsync.*/my/path/to/this/file/$file.*$basedir/$file%, 'Use rsync to copy file');
-            is($last_ssh_commands[-1], "qemu-img create '${basedir}openQA-SUT-1$dev_id.img' -f qcow2 -b '$basedir/$file' $_10gb", 'Used image size <= backingfile size');
+            like($last_ssh_commands[0], qr%^rsync.*/my/path/to/this/file/$file.*$basedir$file%, 'Use rsync to copy file');
```
Very nice, that we got rid of that extra `/`!!

cfconrad reviewed (commented):
Beside the thing that we use the compressed file for "vmkfstools clone" `family vmware backingfile=1 xz file'` 
it looks good to me

cfconrad commented on consoles/sshVirtsh.pm:
```diff
@@ -325,111 +325,167 @@ sub add_interface {
 sub add_disk {
     my ($self, $args) = @_;
+    die('`backingfile` or `cdrom` has been used without a given image file') if (($args->{backingfile} || $args->{cdrom}) && !$args->{file});
+    die('Please specify device id (dev_id)') unless ($args->{dev_id});
```
to the discussion of input validation. You removed now this thing that one of `backingfile`, `create` or `cdrom` is specified.
But if none is specified,
we do `cdrom` for vmware and `create` for kvm. Is this what we expect? actually it looks wired to me, and I would like to have not such behavior!
